File: app/src/main/java/com/volgoak/pokertournament/MainActivity.java
package com.volgoak.pokertournament;

import android.app.ActivityManager;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.aigestudio.wheelpicker.WheelPicker;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.volgoak.pokertournament.admob.AdsManager;

import com.volgoak.pokertournament.admob.ConsentManager;
import com.volgoak.pokertournament.data.Structure;

import com.volgoak.pokertournament.data.StructureProvider;
import com.volgoak.pokertournament.databinding.ActivityMainBinding;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity{

    public static final String TAG = "MainActivity";
    public static final String SAVED_TIME_POSITION = "saved_time";
    public static final String SAVED_BLIND_POSITION = "saved_blind";
    public static final String SAVED_STRUCTURE_POSITION = "saved_structure";

    private ActivityMainBinding mBinding;

    private List<Structure> mStructureList;
    private List<String> mTimeList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        FirebaseAnalytics.getInstance(this);

        //check if BlindsService already started
        boolean isServiceRunning = isTournamentStarted();
        //if a BlindsService started launch TournamentActivity
        if(isServiceRunning){
            Intent intent = new Intent(this, TournamentActivity.class);
            startActivity(intent);
            finish();
        }
        //initialize dataBinder
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        //get strings with time options and setup wheelPicker with time
        mTimeList = new ArrayList<>();
        for(int i = 1; i < 41; i++) mTimeList.add(String.valueOf(i));

        mBinding.wheelRoundTimeMain.setData(mTimeList);
        setupWheel(mBinding.wheelRoundTimeMain);

        //setup structure wheelPicker
        mStructureList = StructureProvider.getStructures(this);
        mBinding.wheelBlindsStructureMain.setData(mStructureList);
        setupWheel(mBinding.wheelBlindsStructureMain);
        mBinding.wheelBlindsStructureMain.setOnItemSelectedListener(new WheelPicker.OnItemSelectedListener() {
            @Override
            public void onItemSelected(WheelPicker picker, Object data, int position) {
                renewBlindsList(position);
            }
        });

        //setup blinds wheel
        renewBlindsList(0);
        setupWheel(mBinding.wheelStartBlindMain);

        //create listener for the start button
        mBinding.btStartMain.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startGame();
            }
        });

        AdsManager.INSTANCE.checkConsent(this);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        int structurePosition = savedInstanceState.getInt(SAVED_STRUCTURE_POSITION, 0);
        mBinding.wheelBlindsStructureMain.setSelectedItemPosition(structurePosition);

        int blindPosition = savedInstanceState.getInt(SAVED_BLIND_POSITION, 0);
        mBinding.wheelStartBlindMain.setSelectedItemPosition(blindPosition);

        int timePosition = savedInstanceState.getInt(SAVED_TIME_POSITION, 0);
        mBinding.wheelRoundTimeMain.setSelectedItemPosition(timePosition);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(SAVED_STRUCTURE_POSITION, mBinding.wheelBlindsStructureMain.getCurrentItemPosition());
        outState.putInt(SAVED_BLIND_POSITION, mBinding.wheelStartBlindMain.getCurrentItemPosition());
        outState.putInt(SAVED_TIME_POSITION, mBinding.wheelRoundTimeMain.getCurrentItemPosition());
    }

    private void startGame(){
        //get selected round time and parse it to millis
        int selectedMinuts = mBinding.wheelRoundTimeMain.getCurrentItemPosition();
        Log.d(TAG, "startGame: selectedPosition " + selectedMinuts);
        String minutesString = mTimeList.get(selectedMinuts);
        Log.d(TAG, "startGame: minutesString = " + minutesString);
        long minutesLong = Long.parseLong(minutesString);
        long roundTime = minutesLong * 60 * 1000;

        int startBlindPosition = mBinding.wheelStartBlindMain.getCurrentItemPosition();

        Structure selectedStructure = mStructureList
                .get(mBinding.wheelBlindsStructureMain.getCurrentItemPosition());

        //Put tournament info into Intent and start service
        Intent intent = new Intent(this, BlindsService.class);
        intent.putExtra(BlindsService.EXTRA_STRUCTURE, selectedStructure);
        intent.putExtra(BlindsService.EXTRA_ROUND_TIME, roundTime);
        intent.putExtra(BlindsService.EXTRA_START_ROUND, --startBlindPosition);
        intent.setAction(BlindsService.START_GAME_ACTION);
        startService(intent);

        //Start tournament activity
        Intent activityIntent = new Intent(this, TournamentActivity.class);
        startActivity(activityIntent);

        Log.d(TAG, "startGame: newVersion");
        //finish activity to avoid returning while tournament in progress
        finish();
    }

    //prepare wheelPicker
    private void setupWheel(WheelPicker wheel){
        wheel.setAtmospheric(true);
        wheel.setItemTextSize(getResources().getDimensionPixelSize(R.dimen.wheel_text_size));
        wheel.setCurved(true);
        wheel.setCyclic(true);
        wheel.setVisibleItemCount(4);
    }

    //renew start blind wheel
    private void renewBlindsList(int position){
        Structure structure = mStructureList.get(position);
        mBinding.wheelStartBlindMain.setData(structure.getBlinds());
    }

    /**
     * Get all system services and check if BlindsService
     * already ran
     * @return true if BlindsService ran
     */
    private boolean isTournamentStarted(){
        ActivityManager activityManager = (ActivityManager) getSystemService(ACTIVITY_SERVICE);
        List<ActivityManager.RunningServiceInfo> servicesInfo = activityManager.getRunningServices(Integer.MAX_VALUE);

        String blindsService = BlindsService.class.getName();
        boolean serviceRunning = false;

        for(int a = 0; a < servicesInfo.size(); a++){
            Log.d(TAG, servicesInfo.get(a).service.getClassName());
            if(blindsService.equals(servicesInfo.get(a).service.getClassName())){
                serviceRunning = true;
                break;
            }
        }
        return serviceRunning;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()){
            case R.id.menu_about :
                Intent intent = new Intent(this, AboutActivity.class);
                startActivity(intent);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/volgoak/pokertournament/MainActivity.kt
package com.volgoak.pokertournament

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.aigestudio.wheelpicker.WheelPicker
import com.google.firebase.analytics.FirebaseAnalytics
import com.volgoak.pokertournament.admob.AdsManager
import com.volgoak.pokertournament.data.Structure
import com.volgoak.pokertournament.data.StructureProvider
import com.volgoak.pokertournament.service.BlindsService
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val SAVED_TIME_POSITION = "saved_time"
        private const val SAVED_BLIND_POSITION = "saved_blind"
        private const val SAVED_STRUCTURE_POSITION = "saved_structure"

        fun getIntent(context: Context) = Intent(context, MainActivity::class.java)
    }

    private var mStructureList: List<Structure>? = null
    private var timeList = listOf<String>()

    /**
     * Get all system services and check if BlindsService
     * already ran
     * @return true if BlindsService ran
     *//*
    private val isTournamentStarted: Boolean
        get() {
            val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            val servicesInfo = activityManager.getRunningServices(Integer.MAX_VALUE)

            val blindsService = BlindsService::class.java.name
            var serviceRunning = false

            for (a in servicesInfo.indices) {
                if (blindsService == servicesInfo[a].service.className) {
                    serviceRunning = true
                    break
                }
            }
            return serviceRunning
        }*/
    //todo check is service already started

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        FirebaseAnalytics.getInstance(this)

        //check if BlindsService already started
        val isServiceRunning = false // todo set real value
        //if a BlindsService started launch TournamentActivity
        if (isServiceRunning) {
            val intent = Intent(this, TournamentActivity::class.java)
            startActivity(intent)
            finish()
        }

        //get strings with time options and setup wheelPicker with time
        timeList = (1..60).toList()
                .map { it.toString() }

        wheelRoundTime.data = timeList
        setupWheel(wheelRoundTime)

        //setup structure wheelPicker
        mStructureList = StructureProvider.getStructures(this)
        wheelBlindsStructure.data = mStructureList!!
        setupWheel(wheelBlindsStructure)
        wheelBlindsStructure.setOnItemSelectedListener { picker, data, position -> renewBlindsList(position) }

        //setup blinds wheel
        renewBlindsList(0)
        setupWheel(wheelStartBlinds)

        //create listener for the start button
        fabStart.setOnClickListener { startGame() }

        AdsManager.checkConsent(this)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)

        val structurePosition = savedInstanceState.getInt(SAVED_STRUCTURE_POSITION, 0)
        wheelBlindsStructure.selectedItemPosition = structurePosition

        val blindPosition = savedInstanceState.getInt(SAVED_BLIND_POSITION, 0)
        wheelStartBlinds.selectedItemPosition = blindPosition

        val timePosition = savedInstanceState.getInt(SAVED_TIME_POSITION, 0)
        wheelRoundTime.selectedItemPosition = timePosition
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(SAVED_STRUCTURE_POSITION, wheelBlindsStructure.currentItemPosition)
        outState.putInt(SAVED_BLIND_POSITION, wheelStartBlinds.currentItemPosition)
        outState.putInt(SAVED_TIME_POSITION, wheelRoundTime.currentItemPosition)
    }

    private fun startGame() {
        //get selected round time and parse it to millis
        val selectedMinuts = wheelRoundTime.currentItemPosition
        val minutesString = timeList!![selectedMinuts]
        val minutesLong = java.lang.Long.parseLong(minutesString)
        val roundTime = minutesLong * 60 * 1000

        val startBlindPosition = wheelStartBlinds.currentItemPosition - 1

        val selectedStructure = mStructureList!![wheelBlindsStructure.currentItemPosition]

        //Put tournament info into Intent and start service
        val intent = BlindsService.getStartGameIntent(
                this,
                selectedStructure,
                roundTime,
                startBlindPosition
        )

        ContextCompat.startForegroundService(this, intent)

        //Start tournament activity
        val activityIntent = Intent(this, TournamentActivity::class.java)
        startActivity(activityIntent)

        //finish activity to avoid returning while tournament in progress
        finish()
    }

    //prepare wheelPicker
    private fun setupWheel(wheel: WheelPicker) {
        wheel.setAtmospheric(true)
        wheel.itemTextSize = resources.getDimensionPixelSize(R.dimen.wheel_text_size)
        wheel.isCurved = true
        wheel.isCyclic = true
        wheel.visibleItemCount = 4
    }

    //renew start blind wheel
    private fun renewBlindsList(position: Int) {
        val structure = mStructureList!![position]
        wheelStartBlinds.data = structure.blinds
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_about -> {
                val intent = Intent(this, AboutActivity::class.java)
                startActivity(intent)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
